lib/tree.rb

```diff
@@ -2,61 +2,210 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n)) in a balanced tree, where n is the number of nodes in the tree. O(n) in an unbalanced tree.
+  # Space Complexity: O(n), where n is the number of nodes in the tree
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    if !@root
+      @root = new_node
+    else
+      current = @root
+      move_and_add(new_node, current)
+      return
+    end
+  end
+
+  def move_and_add(new, current)
+    if new.key < current.key
+      if !current.left
+        current.left = new
+        return
+      else
+        move_and_add(new, current.left)
+      end
+    elsif new.key > current.key
+      if !current.right
+        current.right = new
+        return
+      else
+        move_and_add(new, current.right)
+      end
+    elsif new.key == current.key
+      return
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n)) in a balanced tree, where n is the number of nodes in the tree. O(n) in an unbalanced tree.
+  # Space Complexity: O(m), where m is the height of the tree
   def find(key)
-    raise NotImplementedError
+    if !@root
+      return nil
+    end
+    current = @root
+
+    return move_and_find(key, current)
+  end
+
+  def move_and_find(key, current)
```

Comment:
I like these two helper methods, but they look like they could use a refactor. Just a little messy, this is a minor issue.